src/ImageCore.jl

```diff
@@ -30,6 +29,15 @@ Color1Array{C<:Color1,N} = AbstractArray{C,N}
 const RRArray{To,From,N,M,P} = Base.ReshapedArray{To,N,Base.ReinterpretArray{To,M,From,P}}
 const RGArray = Union{Base.ReinterpretArray{<:AbstractGray,M,<:Number,P}, Base.ReinterpretArray{<:Number,M,<:AbstractGray,P}} where {M,P}
 
+# delibrately not export these constants to enable extensibility for downstream packages
+const NumberLike{T<:Number} = Union{T,AbstractGray{T}}
+const RealLike{T<:Real} = NumberLike{T}
+const FloatLike{T<:AbstractFloat} = RealLike{T}
+const FractionalLike{T<:Union{FixedPoint, AbstractFloat}} = RealLike{T}
+const GrayLike{T<:Union{Bool, FixedPoint, AbstractFloat}} = RealLike{T}
+const GrayImage{N, BT, T<:GrayLike{BT}} = AbstractArray{<:GrayLike{T}, N}
```

Comment:
I should have noticed this earlier, sorry:

```julia
julia> getBT(img::GrayImage{N,BT,T}) where {N,BT,T} = BT
getBT (generic function with 1 method)

julia> getT(img::GrayImage{N,BT,T}) where {N,BT,T} = T
getT (generic function with 1 method)

julia> img = Gray.(rand(2,2))
2×2 Array{Gray{Float64},2} with eltype Gray{Float64}:
 Gray{Float64}(0.524819)  Gray{Float64}(0.508767)
 Gray{Float64}(0.939333)  Gray{Float64}(0.258288)

julia> getBT(img)
ERROR: UndefVarError: BT not defined
Stacktrace:
 [1] getBT(::Array{Gray{Float64},2}) at ./REPL[23]:1
 [2] top-level scope at none:0

julia> getT(img)
Float64
```

Reply:
wow that's a surprise! I don't know why this would happen, any hints?

But this might not be the problem WRT the dispatch purpose, which is covered in the tests.

```julia
foo(img::Gray2dImage) = "Generic"
foo(img::Gray2dImage{<:AbstractFloat}) = "AbstractFloat"
foo(img::Gray2dImage{<:FixedPoint}) = "FixedPoint"
@test foo(rand(Bool, sz)) == "Generic"
@test foo(rand(Gray{Bool}, sz)) == "Generic"
@test foo(rand(Gray{Float32}, sz)) == "AbstractFloat"
@test foo(rand(Float32, sz)) == "AbstractFloat"
@test foo(rand(Gray{N0f8}, sz)) == "FixedPoint"
@test foo(rand(N0f8, sz)) == "FixedPoint"
```

We can always use `getBT(img::GrayImage) == eltype(eltype(img))` 😄

Reply:
I think it's a bad idea to introduce parameters to types that have no way of being set. Is there anything wrong with `const GrayImage{T<:GrayLike,N} = AbstractArray{T,N}`?

```julia
julia> foo(img::GrayImage{T}) where T<:FloatLike = "floatlike gray image"
foo (generic function with 1 method)

julia> foo(img)  = "some other thing"
foo (generic function with 2 methods)

julia> foo(rand(2,2))
"floatlike gray image"

julia> foo(Gray.(rand(2,2)))
"floatlike gray image"

julia> foo(Gray.(rand(Bool,2,2)))
"some other thing"
```

Reply:
That's it!
I was making things too complicated to confuse myself.